drivers/ws281x/systick_gpio_ll.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2024 Marian Buschsieweke
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ws281x
+ *
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the WS281x abstraction based on GPIO_LL and timers
+ *
+ * @author      Marian Buschsieweke <marian.buschsieweke@posteo.net>
+ *
+ * @}
+ */
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "clk.h"
+#include "cpu.h"
+#include "irq.h"
+#include "macros/math.h"
+#include "periph/gpio_ll.h"
+#include "time_units.h"
+
+#include "ws281x.h"
+#include "ws281x_params.h"
+#include "ws281x_constants.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static void _systick_start(uint32_t ticks)
+{
+    /* disable SysTick, clear value */
+    SysTick->CTRL = 0;
+    SysTick->VAL = 0;
+    /* prepare value in re-load register */
+    SysTick->LOAD = ticks;
+    /* start and wait for the load value to be applied */
+    SysTick->CTRL = SysTick_CTRL_CLKSOURCE_Msk | SysTick_CTRL_ENABLE_Msk;
+    while (SysTick->VAL == 0) { /* Wait for SysTick to start and spin */ }
+}
+
+static void _systick_wait(void)
+{
+    while (!(SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk)) { /* busy wait */ }
+}
+
+void ws281x_write_buffer(ws281x_t *dev, const void *buf, size_t size)
+{
+    assert(dev);
+
+    /* the high time for one can be as high as 5 seconds in practise, so
+     * rather be on the high side by adding a few CPU cycles. */
+    const uint32_t ticks_one = DIV_ROUND_UP((uint64_t)WS281X_T_DATA_ONE_NS * (uint64_t)coreclk(), NS_PER_SEC) + 16;
+    /* the low time should rather be on the short side, so rounding down */
+    const uint32_t ticks_zero = (uint64_t)(WS281X_T_DATA_ZERO_NS - 50) * (uint64_t)coreclk() / NS_PER_SEC;
```

Comment:
For 48 MHz this is:

(325 - 50) * 48 * 10^6 / 10^9 = 13.2

Reply:
so 13 cyles for one wait ( which has (a load, a compare and a jump back if bit 16 isn't set)  i think this is plenty of time to do this multiple times

the short wait has exact the same number of wait as with your code but it does not do a setup in the low period

i thought you where talking about 13cycle for the 1 µsecond